library/alloc/src/collections/vec_deque/iter.rs

```diff
@@ -142,28 +112,36 @@ impl<'a, T> Iterator for Iter<'a, T> {
 impl<'a, T> DoubleEndedIterator for Iter<'a, T> {
     #[inline]
     fn next_back(&mut self) -> Option<&'a T> {
-        if self.tail == self.head {
-            return None;
+        match self.i2.next_back() {
+            Some(val) => Some(val),
+            None => {
+                // most of the time, the iterator will either always
+                // call next(), or always call next_back(). By swapping
+                // the iterators once the first one is empty, we ensure
+                // that the first branch is taken as often as possible,
+                // without sacrificing correctness, as i2 is empty anyways
+                mem::swap(&mut self.i1, &mut self.i2);
```

Comment:
pondering: at this point, the `self.i1` is not useful, so I wonder if *swapping* is really the best thing here.

It would also be legal to do `self.i2 = mem::replace(&mut self.i1, [].iter());`, right?  Dunno if that's really better, though.  I guess it's definitely not something we need to figure out in this PR.

Reply:
btw, you can `mem::take` iterators now since they implement Default :)